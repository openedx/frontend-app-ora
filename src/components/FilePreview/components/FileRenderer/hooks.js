import { useState } from 'react';
import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
