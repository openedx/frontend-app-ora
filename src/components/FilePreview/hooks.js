import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from 'components/FilePreview/BaseRenderers';

import messages from './messages';

export const ErrorStatuses = StrictDict({
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
});

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
});

/**
 * Config data
 */
export const RENDERERS = {
  [FileTypes.pdf]: PDFRenderer,
  [FileTypes.jpg]: ImageRenderer,
  [FileTypes.jpeg]: ImageRenderer,
  [FileTypes.bmp]: ImageRenderer,
  [FileTypes.png]: ImageRenderer,
  [FileTypes.txt]: TXTRenderer,
  [FileTypes.gif]: ImageRenderer,
  [FileTypes.jfif]: ImageRenderer,
  [FileTypes.pjpeg]: ImageRenderer,
  [FileTypes.pjp]: ImageRenderer,
  [FileTypes.svg]: ImageRenderer,
};

export const SUPPORTED_TYPES = Object.keys(RENDERERS);

export const ERROR_STATUSES = {
  [ErrorStatuses.notFound]: messages.fileNotFoundError,
  [ErrorStatuses.serverError]: messages.unknownError,
};

/**
 * Util methods and transforms
 */
export const getFileType = (fileName) => fileName.split('.').pop()?.toLowerCase();
export const isSupported = (file) => SUPPORTED_TYPES.includes(
  getFileType(file.name),
);

export const stateKeys = StrictDict({
  errorStatus: 'errorStatus',
  isLoading: 'isLoading',
});

/**
 * component hooks
 */
export const renderHooks = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useKeyedState(stateKeys.errorStatus, null);
  const [isLoading, setIsLoading] = useKeyedState(stateKeys.isLoading, true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    ERROR_STATUSES[errorStatus] || ERROR_STATUSES[ErrorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = RENDERERS[getFileType(file.name)];
  const rendererProps = {
    fileName: file.name,
    url: file.downloadUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
