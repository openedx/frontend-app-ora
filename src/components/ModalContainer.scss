@import '@edx/brand/paragon/variables.scss';
@import '@edx/paragon/scss/core/core.scss';
@import '@edx/brand/paragon/overrides.scss';

.ora-modal-body {
  min-height: calc(100vh - 20px - 64px - (map-get($spacers, 2) * 8));
  height: 100%;
}

.exam-timer {
  max-width: $max-width-lg;
  margin: auto;
  button {
    display: none;
  }
}
