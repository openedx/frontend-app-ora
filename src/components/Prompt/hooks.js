import { useState } from 'react';

const usePromptHooks = () => {
  const [open, setOpen] = useState(true);

  const toggleOpen = () => setOpen(!open);

  return {
    open,
    toggleOpen,
  };
};

export default usePromptHooks;
