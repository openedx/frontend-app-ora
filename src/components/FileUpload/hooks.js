import { useCallback } from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import { useDownloadFiles } from 'hooks/app';

export const stateKeys = StrictDict({
  shouldShowError: 'shouldShowError',
  isModalOpen: 'isModalOpen',
  uploadArgs: 'uploadArgs',
  description: 'description',
});

export const useUploadConfirmModalHooks = ({
  file,
  closeHandler,
  uploadHandler,
}) => {
  const [description, setDescription] = useKeyedState(stateKeys.description, '');
  const [shouldShowError, setShouldShowError] = useKeyedState(
    stateKeys.shouldShowError,
    false,
  );

  const confirmUploadClickHandler = () => {
    // console.log({ confirmUploadClick: { description } });
    if (description !== '') {
      uploadHandler(file, description);
    } else {
      setShouldShowError(true);
    }
  };

  const exitHandler = () => {
    setShouldShowError(false);
    setDescription('');
    closeHandler();
  };

  // Modifying pointer of file object. This is not a good practice.
  // eslint-disable-next-line no-param-reassign, no-return-assign
  const onFileDescriptionChange = (event) => setDescription(event.target.value);

  return {
    shouldShowError,
    confirmUploadClickHandler,
    exitHandler,
    onFileDescriptionChange,
  };
};

export const useFileUploadHooks = ({ onFileUploaded }) => {
  const [uploadArgs, setUploadArgs] = useKeyedState(stateKeys.uploadArgs, {});
  const [isModalOpen, setIsModalOpen] = useKeyedState(
    stateKeys.isModalOpen,
    false,
  );

  const confirmUpload = useCallback(async (file, description) => {
    // console.log({ confirmUpload: { file, description } });
    setIsModalOpen(false);
    if (onFileUploaded) {
      // console.log({ uploadArgs });
      await onFileUploaded({ ...uploadArgs, description });
    }
    setUploadArgs({});
  }, [uploadArgs, onFileUploaded, setIsModalOpen, setUploadArgs]);

  const closeUploadModal = useCallback(() => {
    setIsModalOpen(false);
    setUploadArgs({});
  }, [setIsModalOpen, setUploadArgs]);

  const onProcessUpload = useCallback(
    ({ fileData, handleError, requestConfig }) => {
      setIsModalOpen(true);
      setUploadArgs({ fileData, handleError, requestConfig });
    },
    [setIsModalOpen, setUploadArgs],
  );

  return {
    isModalOpen,
    uploadArgs,
    confirmUpload,
    closeUploadModal,
    onProcessUpload,
  };
};

export const useFileDownloadHooks = ({ files, zipFileName }) => {
  const downloadFileMation = useDownloadFiles();
  const downloadFiles = () => downloadFileMation.mutate({
    files,
    zipFileName,
  });
  return {
    downloadFiles,
    status: downloadFileMation.status,
  };
};

export default {
  useUploadConfirmModalHooks,
  useFileUploadHooks,
  useFileDownloadHooks,
};
