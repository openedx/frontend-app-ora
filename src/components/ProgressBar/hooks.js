import { useParams } from 'react-router-dom';
import { useIsEmbedded, useViewStep } from 'hooks';
import {
  useEffectiveGrade,
  useGlobalState,
} from 'data/services/lms/hooks/selectors';
import {
  stepRoutes,
  stepStates,
} from 'data/services/lms/constants';

export const useProgressStepData = ({ step, canRevisit = false }) => {
  const { xblockId, courseId } = useParams();
  const isEmbedded = useIsEmbedded();
  const viewStep = useViewStep();
  const { stepState } = useGlobalState({ step });

  const href = `/${stepRoutes[step]}${isEmbedded ? '/embedded' : ''}/${courseId}/${xblockId}`;
  const isActive = viewStep === step;
  const isEnabled = (
    isActive
    || (stepState === stepStates.inProgress)
    || (canRevisit && stepState === stepStates.done)
  );
  const myGrade = useEffectiveGrade()?.stepScore;
  return {
    href,
    isEnabled,
    isActive,
    isComplete: stepState === stepStates.done,
    inProgress: stepState === stepStates.inProgress,
    isPastDue: stepState === stepStates.closed,
    myGrade,
    // myGrade: { earned: 8, possible: 10 },
    // isPastDue: step === 'self',
  };
};

export default { useProgressStepData };
