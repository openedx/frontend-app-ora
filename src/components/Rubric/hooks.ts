import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';
import { usePageData, useRubricConfig } from 'data/services/lms/hooks/selectors';
import { submitRubric } from 'data/services/lms/hooks/actions';
import { RubricHookData } from './types';

export const stateKeys = StrictDict({
  rubric: 'rubric',
  overallFeedback: 'overallFeedback',
});

export const useRubricData = ({ isGrading }): RubricHookData => {
  const data = usePageData();
  const { criteria, feedbackConfig } = useRubricConfig();

  const [rubricData, setRubricData] = useKeyedState(stateKeys.rubric, data.rubric);
  const [overallFeedback, setOverallFeedback] = useKeyedState(
    stateKeys.overallFeedback, data.rubric.overallFeedback
  );
  const submitRubricMutation = submitRubric();

  const onOverallFeedbackChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setOverallFeedback(e.target.value);

  const onSubmit = () => {
    submitRubricMutation.mutate({
      overallFeedback,
      optionsSelected: rubricData.optionsSelected,
      criterionFeedback: rubricData.criterionFeedback,
    } as any);
  };

  return {
    rubricData,
    setRubricData,
    criteria: criteria.map((criterion) => ({
      ...criterion,
      optionsValue: rubricData.optionsSelected[criterion.name],
      optionsIsInvalid: false,
      optionsOnChange: (e: React.ChangeEvent<HTMLInputElement>) => {
        setRubricData({
          ...rubricData,
          optionsSelected: {
            ...rubricData.optionsSelected,
            [criterion.name]: e.target.value,
          },
        });
      },

      feedbackValue: rubricData.criterionFeedback[criterion.name],
      feedbackIsInvalid: false,
      feedbackOnChange: (e: React.ChangeEvent<HTMLInputElement>) => {
        setRubricData({
          ...rubricData,
          criterionFeedback: {
            ...rubricData.criterionFeedback,
            [criterion.name]: e.target.value,
          },
        });
      },
    })),
    onSubmit,
    submitStatus: submitRubricMutation.status,

    // overall feedback
    overallFeedback,
    onOverallFeedbackChange,
    overallFeedbackDisabled: !isGrading,
    overallFeedbackIsInvalid: false,
    overallFeedbackPrompt: feedbackConfig.defaultText,
  };
};

export default useRubricData;
