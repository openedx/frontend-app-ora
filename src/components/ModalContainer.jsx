import React from 'react';
import PropTypes from 'prop-types';
import { useORAConfigData } from 'hooks/app';

import ProgressBar from 'components/ProgressBar';

/* The purpose of this component is to wrap views with a header/footer for situations
 * where we need to run non-embedded. It is a replicated style of FullScreenModal from
 * paragon. The reason we use this instead of FullScreenModal is because FullScreenModal
 * is very opinionated on other components that it uses on top of it. Since we are not
 * using the modality of the component, it would be better to just replicate the style
 * instead of using the component.
 */

const ModalContainer = ({ children }) => {
  const { title } = useORAConfigData();
  return (
    <div className="h-100">
      <div className="sticky-top bg-white">
        <h3 className="bg-dark text-white p-3 m-0">{title}</h3>
        <ProgressBar className="px-2 shadow-sm" />
      </div>
      <div className="content-body bg-light-300 p-4">
        {children}
      </div>
    </div>
  );
};
ModalContainer.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ModalContainer;
