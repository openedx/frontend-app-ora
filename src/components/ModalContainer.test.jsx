import { shallow } from '@edx/react-unit-test-utils';

import { useORAConfigData } from 'hooks/app';
import { useCloseModalAction } from 'hooks/actions';
import ModalContainer from './ModalContainer';

jest.mock('hooks/app', () => ({
  useORAConfigData: jest.fn(),
}));
jest.mock('hooks/actions', () => ({
  useCloseModalAction: jest.fn(),
}));
jest.mock('components/ConfirmDialog', () => 'ConfirmDialog');
jest.mock('components/ProgressBar', () => 'ProgressBar');
jest.mock('react-router', () => ({
  useParams: jest.fn().mockReturnValue({ courseId: 'courseId' }),
}));
jest.mock('@edx/frontend-lib-special-exams', () => ({
  OuterExamTimer: () => 'OuterExamTimer',
}));

describe('<ModalContainer />', () => {
  useORAConfigData.mockReturnValue({ title: 'title' });
  useCloseModalAction.mockReturnValue({
    confirmProps: {
      abc: 'def',
    },
    action: { onClick: jest.fn().mockName('closeModalAction') },
  });
  const renderComponent = () => shallow(
    <ModalContainer>
      <div>children</div>
    </ModalContainer>,
  );

  it('render default', () => {
    const wrapper = renderComponent();
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('ConfirmDialog')).toHaveLength(1);
  });

  it('render without confirmProps', () => {
    useCloseModalAction.mockReturnValue({
      confirmProps: null,
      action: { onClick: jest.fn().mockName('closeModalAction') },
    });
    const wrapper = renderComponent();
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('ConfirmDialog')).toHaveLength(0);
  });
});
