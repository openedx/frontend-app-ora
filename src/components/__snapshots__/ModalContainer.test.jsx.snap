// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalContainer /> render default 1`] = `
<div>
  <ConfirmDialog
    abc="def"
  />
  <div
    className="sticky-top bg-white"
  >
    <h3
      className="w-100 bg-dark text-white p-3 m-0"
    >
      title
    </h3>
    <div
      style={
        Object {
          "insetInlineEnd": "1rem",
          "position": "absolute",
          "right": "0px",
          "top": ".625rem",
        }
      }
    >
      <Button
        className="text-white"
        onClick={[MockFunction closeModalAction]}
      >
        <Icon
          alt="Close"
        />
      </Button>
    </div>
    <ProgressBar
      className="px-2 shadow-sm"
    />
  </div>
  <div
    className="ora-modal-body content-body bg-light-300 p-4"
  >
    <div>
      children
    </div>
  </div>
</div>
`;

exports[`<ModalContainer /> render without confirmProps 1`] = `
<div>
  <div
    className="sticky-top bg-white"
  >
    <h3
      className="w-100 bg-dark text-white p-3 m-0"
    >
      title
    </h3>
    <div
      style={
        Object {
          "insetInlineEnd": "1rem",
          "position": "absolute",
          "right": "0px",
          "top": ".625rem",
        }
      }
    >
      <Button
        className="text-white"
        onClick={[MockFunction closeModalAction]}
      >
        <Icon
          alt="Close"
        />
      </Button>
    </div>
    <ProgressBar
      className="px-2 shadow-sm"
    />
  </div>
  <div
    className="ora-modal-body content-body bg-light-300 p-4"
  >
    <div>
      children
    </div>
  </div>
</div>
`;
